cmd/openshift-install/gather.go

```diff
@@ -0,0 +1,235 @@
+	assetstore "github.com/openshift/installer/pkg/asset/store"
+	"github.com/openshift/installer/pkg/terraform"
+	"github.com/openshift/installer/pkg/types"
+	awstypes "github.com/openshift/installer/pkg/types/aws"
+	libvirttypes "github.com/openshift/installer/pkg/types/libvirt"
+	openstacktypes "github.com/openshift/installer/pkg/types/openstack"
+)
+
+func newGatherCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "gather",
+		Short: "Gather debugging data for a given installation failure",
+		Long: `Gather debugging data for a given installation failure.
+
+When installation for Openshift cluster fails, gathering all the data useful for debugging can
+become a difficult task. This command helps users to collect the most relevant information that can be used
+to debug the installation failures`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cmd.Help()
+		},
+	}
+	cmd.AddCommand(newGatherBootstrapCmd())
+	return cmd
+}
+
+var (
+	gatherBootstrapOpts struct {
+		bootstrap string
+		masters   []string
+	}
+)
+
+func newGatherBootstrapCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "bootstrap",
+		Short: "Gather debugging data for a failing to bootstrap control plane",
+		Args:  cobra.ExactArgs(0),
+		Run: func(_ *cobra.Command, _ []string) {
+			cleanup := setupFileHook(rootOpts.dir)
+			defer cleanup()
+			err := runGatherBootstrapCmd(rootOpts.dir)
+			if err != nil {
+				logrus.Fatal(err)
+			}
+		},
+	}
+	cmd.PersistentFlags().StringVar(&gatherBootstrapOpts.bootstrap, "bootstrap", "", "")
+	cmd.PersistentFlags().StringArrayVar(&gatherBootstrapOpts.masters, "master", []string{}, "")
+	return cmd
+}
+
+func runGatherBootstrapCmd(directory string) error {
+	tfStateFilePath := filepath.Join(directory, terraform.StateFileName)
+	_, err := os.Stat(tfStateFilePath)
+	if os.IsNotExist(err) {
+		return unSupportedPlatformGather()
+	}
+	if err != nil {
+		return err
+	}
+
+	assetStore, err := assetstore.NewStore(directory)
+	if err != nil {
+		return errors.Wrap(err, "failed to create asset store")
+	}
+
+	config := &installconfig.InstallConfig{}
+	if err := assetStore.Fetch(config); err != nil {
+		return errors.Wrapf(err, "failed to fetch %s", config.Name())
+	}
+
+	sfRaw, err := ioutil.ReadFile(tfStateFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read %q", tfStateFilePath)
+	}
+
+	var tfstate terraformState
+	if err := json.Unmarshal(sfRaw, &tfstate); err != nil {
+		return errors.Wrapf(err, "failed to unmarshal %q", tfStateFilePath)
+	}
+
+	bootstrap, masters, err := extractHostAddresses(config.Config, tfstate)
+	if err != nil {
+		if err2, ok := err.(errUnSupportedGatherPlatform); ok {
+			logrus.Error(err2)
+			return unSupportedPlatformGather()
+		}
+		return errors.Wrapf(err, "failed to get bootstrap and control plane host addresses from %q", tfStateFilePath)
+	}
+
+	logrus.Info("Use the following commands to gather logs from the cluster")
```

Comment:
can we move this to https://github.com/openshift/installer/pull/1627/files#diff-be4790dcfa83ca14e7942b4c3b50c5f9R118 ??

Reply:
Done.